File: ios/polar/Sources/polar/FeatureReadinessCache.swift
import Foundation
import PolarBleSdk

/// Thread-safe record of definitive feature-ready callbacks seen by the plugin.
final class FeatureReadinessCache {
  private let lock = NSLock()
  private var featuresByDevice: [String: Set<PolarBleSdkFeature>] = [:]

  func record(_ identifier: String, feature: PolarBleSdkFeature) {
    lock.lock()
    defer { lock.unlock() }
    featuresByDevice[identifier, default: []].insert(feature)
  }

  func isReady(
    _ identifier: String,
    feature: PolarBleSdkFeature,
    nativeReady: Bool
  ) -> Bool {
    if nativeReady { return true }
    lock.lock()
    defer { lock.unlock() }
    return featuresByDevice[identifier]?.contains(feature) == true
  }

  func clear(_ identifier: String) {
    lock.lock()
    defer { lock.unlock() }
    featuresByDevice.removeValue(forKey: identifier)
  }

  func clearAll() {
    lock.lock()
    defer { lock.unlock() }
    featuresByDevice.removeAll()
  }
}

File: android/src/main/kotlin/dev/rexios/polar/FeatureReadinessCache.kt
package dev.rexios.polar

import com.polar.sdk.api.PolarBleApi.PolarBleSdkFeature

/** Thread-safe record of definitive feature-ready callbacks seen by the plugin. */
internal class FeatureReadinessCache {
    private val lock = Any()
    private val featuresByDevice = mutableMapOf<String, MutableSet<PolarBleSdkFeature>>()

    fun record(
        identifier: String,
        feature: PolarBleSdkFeature,
    ): Unit = synchronized(lock) {
        featuresByDevice.getOrPut(identifier) { mutableSetOf() }.add(feature)
        Unit
    }

    fun recordReadiness(
        identifier: String,
        ready: List<PolarBleSdkFeature>,
        unavailable: List<PolarBleSdkFeature>,
    ): Unit = synchronized(lock) {
        val features = featuresByDevice.getOrPut(identifier) { mutableSetOf() }
        features.removeAll(unavailable.toSet())
        features.addAll(ready)
        if (features.isEmpty()) featuresByDevice.remove(identifier)
        Unit
    }

    fun isReady(
        identifier: String,
        feature: PolarBleSdkFeature,
        nativeReady: Boolean,
    ): Boolean = nativeReady || synchronized(lock) {
        featuresByDevice[identifier]?.contains(feature) == true
    }

    fun clear(identifier: String) = synchronized(lock) {
        featuresByDevice.remove(identifier)
        Unit
    }

    fun clearAll() = synchronized(lock) {
        featuresByDevice.clear()
    }
}
